File: frontend/src/types/chemical.ts
/**
 * Shared TypeScript types mirroring the database schema (db/schema.sql).
 * Kept in sync by hand — update both when the schema changes.
 */

export type UserRole = "admin" | "manager" | "viewer";

export type UploadJobStatus =
  | "queued"
  | "processing"
  | "paused"
  | "done"
  | "failed"
  | "cancelled";

/**
 * One document in the DB-worker pipeline (backend documents/pages tables).
 * Replaces the old in-memory UploadJob — progress is derived from the
 * documents/pages status machine, so it survives reloads and backend restarts.
 */
export type DocumentStatusValue =
  // Uploaded and durable, but deliberately not claimable: waits for the user to
  // press "Start processing". Nothing costs an AI call until it leaves this state.
  | "staged"
  | "pending"
  | "splitting"
  | "split"
  | "extracting"
  | "paused"
  | "done"
  | "failed"
  | "cancelled";

export interface DocumentStatus {
  id: string;
  content_hash: string;
  filename: string;
  status: DocumentStatusValue;
  page_count: number;
  pages_done: number;
  product_count: number;
  company_name: string | null;
  duplicate: boolean; // skipped because this exact PDF was already processed
  error: string | null;
  created_at: string;
}

/** Date-range filter for the upload history. */
export type UploadRange = "week" | "month" | "year" | "all";

/** Server-paginated upload-history response. */
export interface UploadHistoryResponse {
  items: UploadHistoryItem[];
  count: number; // total matching uploads across ALL pages
  page: number;
  page_size: number;
  total_pages: number;
}

/** One processed-upload record shown in the upload history / admin audit. */
export interface UploadHistoryItem {
  content_hash: string;
  filename: string;
  company_name: string;
  company_name_en: string;
  product_count: number;
  uploaded_by: string | null;
  uploaded_by_email: string | null; // populated only in the admin-wide view
  created_at: string;
}

/** The listings a single upload produced ("which ones were added"). */
export interface DocumentListings {
  content_hash: string;
  filename: string;
  listings: Listing[];
}

/** Aggregate counts for the admin dashboard (GET /admin/dashboard/summary). */
/** Three disjoint listing-quality slices that sum to total_listings. */
export interface StatusDistribution {
  complete: number;
  needs_review: number;
  missing_price: number;
}

/** One bar in the dashboard's "Top suppliers by listings" panel. */
export interface SupplierListingCount {
  company_id: number;
  name: string;
  count: number;
}

export interface DashboardSummary {
  total_listings: number;
  total_suppliers: number;
  uploads_last_7d: number;
  needs_review: number;
  status_distribution: StatusDistribution;
  top_suppliers: SupplierListingCount[];
}

/** One row on the admin user-management page. */
export interface ManagedUser {
  id: string;
  email: string | null;
  role: UserRole;
  created_at: string | null;
  last_sign_in_at: string | null;
}

/** Admin/manager PATCH /listings/{id} body — only the fields to change. */
export interface ListingUpdate {
  name_raw?: string;
  name_en?: string;
  cas_number?: string | null;
  price?: number | null;
  currency?: string | null;
  purity?: string | null;
  needs_review?: boolean;
  /** Full replacement of the technical-details object (reference block kept). */
  details?: Record<string, unknown> | null;
}

/** One supplier in the directory (GET /suppliers). */
export interface Supplier {
  id: number;
  company_name: string;
  company_name_en: string | null;
  email: string | null;
  contact_number: string | null;
  /** Distinct websites printed on this supplier's brochures (from listings). */
  websites: string[];
  listing_count: number;
  created_at: string;
}

/** Server-paginated suppliers directory response. */
export interface SuppliersResponse {
  items: Supplier[];
  count: number; // total suppliers across ALL pages
  page: number;
  page_size: number;
  total_pages: number;
}

/**
 * A flagged listing in the review queue (GET /admin/review) — a Listing plus
 * the upload that produced it (null for uploads predating the ledger column).
 */
export interface ReviewItem extends Listing {
  source_content_hash: string | null;
  source_filename: string | null;
}

/** Server-paginated review queue response. */
export interface ReviewQueueResponse {
  items: ReviewItem[];
  count: number; // total flagged listings across ALL pages
  page: number;
  page_size: number;
  total_pages: number;
}

/** One activity-log entry (GET /admin/audit): who did what, when. */
export interface AuditEntry {
  id: number;
  actor: string | null;
  actor_email: string | null;
  action:
    | "upload"
    | "update_listing"
    | "delete_listing"
    | "bulk_delete_listings"
    | "undo_upload"
    | string;
  details: Record<string, unknown>;
  created_at: string;
}

/** Server-paginated activity log response. */
export interface AuditLogResponse {
  items: AuditEntry[];
  count: number;
  page: number;
  page_size: number;
  total_pages: number;
}

/** One typeahead suggestion for the search box. */
export interface Suggestion {
  type: "chemical" | "supplier";
  label: string; // display text, and what fills the search box when picked
  cas_number: string | null; // set for chemical suggestions when known
  count: number; // how many listings back this suggestion
}

/** A canonical chemical, deduplicated across suppliers. */
export interface Chemical {
  id: string;
  cas_number: string | null;
  name_en: string;
  created_at: string;
}

/** One supplier's listing of a product, as extracted from a brochure. */
export interface Listing {
  id: string;
  chemical_id: string | null;
  company_id: number | null;
  company_name: string; // supplier name as printed (original language)
  company_name_en: string; // English translation of the supplier name
  name_raw: string; // exactly as printed, in the brochure's own language/script
  name_en: string; // English translation for consistent search/sort/dedup
  cas_number: string | null;
  price: number | null;
  currency: string | null;
  /** Price normalized to USD (stored server-side; drives price sorting). */
  price_usd?: number | null;
  /** PKR display value, computed from the current rate at read time. */
  price_pkr?: number | null;
  purity: string | null;
  needs_review: boolean;
  created_at: string;
  /**
   * Flexible per-listing technical details (flash point, hazard class,
   * storage, MOQ, ...) — keys vary per brochure. Only populated by the
   * single-listing detail endpoint; search rows omit it.
   */
  details?: Record<string, unknown> | null;
  /** Website printed on the brochure (the company's GENERAL site). */
  company_website?: string | null;
}

/* ------------------------------------------------------------------ */
/* House knowledge — curated notes managers write from the Inspector   */
/* ------------------------------------------------------------------ */

/** Did the swap work? `avoid` records one that was tried and failed. */
export type SubstitutionVerdict = "works" | "conditional" | "avoid";

export type RegulatoryStatus =
  | "banned"
  | "restricted"
  | "phase_out"
  | "permitted"
  /** Suspected but unconfirmed — kept distinct so it can't harden into a fact. */
  | "unclear";

/**
 * "We used B in place of A, in this context." BosTech's own judgement, and the
 * one source the assistant ranks above its own chemistry.
 */
export interface SubstitutionNote {
  id: string;
  from_chemical_id: string;
  from_name: string | null;
  from_cas: string | null;
  /** Null when the substitute isn't in the catalog — a sourcing lead, not a gap. */
  to_chemical_id: string | null;
  to_name: string;
  to_cas: string | null;
  verdict: SubstitutionVerdict;
  /** Free text, and the point of the note: "GCC floor coatings, summer cure". */
  context: string;
  author_id: string | null;
  author_email: string | null;
  created_at: string;
}

/** A jurisdiction-scoped, dated regulatory status recorded by a manager. */
export interface RegulatoryNote {
  id: string;
  chemical_id: string;
  chemical_name: string | null;
  chemical_cas: string | null;
  jurisdiction: string;
  status: RegulatoryStatus;
  effective_date: string | null;
  note: string;
  source_url: string | null;
  author_id: string | null;
  author_email: string | null;
  created_at: string;
}

/** Everything the house knows about one chemical. */
export interface HouseNotes {
  substitutions: SubstitutionNote[];
  regulatory: RegulatoryNote[];
}

/** A single admin-configurable setting. */
export interface AppSetting {
  key: string;
  value: string;
  is_secret: boolean;
  category: string;
  label: string;
  description: string;
  updated_at: string | null;
}

export interface SettingsCategory {
  key: string;
  label: string;
}

export interface SettingsResponse {
  settings: AppSetting[];
  categories: SettingsCategory[];
}

export interface SystemStatus {
  workers_active: number;
  documents_processing: number;
  total_listings: number;
  total_chemicals: number;
  embeddings_indexed: number;
  extraction_provider: string;
  chat_enabled: boolean;
  embeddings_enabled: boolean;
}

export interface TestKeyResult {
  ok: boolean;
  message: string;
}

File: backend/app/schemas/chemical.py
"""
Pydantic request/response models.

These also validate the JSON Claude returns before it touches the database —
malformed extractions are rejected/logged rather than inserted. Fully defined
in Phase 4; the shapes below are the target.
"""

from typing import Any, Literal

from pydantic import BaseModel, Field, field_validator


class ExtractedProduct(BaseModel):
    """One product row extracted from a brochure."""

    name_raw: str  # exactly as printed, in the brochure's own language/script
    name_en: str  # English translation, always present for search/dedup
    cas_number: str | None = None
    price: float | None = None
    currency: str | None = None
    purity: str | None = None
    # Anything ELSE the brochure prints for this product (flash point, hazard
    # class, storage, MOQ, ...). Keys vary per brochure; absent keys are
    # omitted by the prompt rather than null-filled.
    details: dict[str, Any] | None = None


class ExtractionResult(BaseModel):
    """Full result for one brochure: a company plus its products."""

    company_name: str  # supplier name exactly as printed (any language/script)
    company_name_en: str | None = None  # English translation of the supplier name
    # Website URL if one is printed on the brochure (the company's general
    # site — never a guessed/constructed product link).
    company_website: str | None = None
    # Contact details ONLY if printed on the brochure (never guessed) — they
    # feed the suppliers directory via resolve_company.
    company_email: str | None = None
    company_phone: str | None = None
    products: list[ExtractedProduct]


class ListingOut(BaseModel):
    """One listing row returned by the search endpoint (viewer-safe fields)."""

    id: str
    chemical_id: str | None = None
    company_id: int | None = None
    company_name: str = ""
    company_name_en: str = ""
    name_raw: str
    name_en: str
    cas_number: str | None = None
    price: float | None = None
    currency: str | None = None
    # Cross-currency comparison: price normalized to USD (stored), plus a
    # PKR display value computed from the current exchange rate at read time.
    price_usd: float | None = None
    price_pkr: float | None = None
    purity: str | None = None
    needs_review: bool = False
    created_at: str
    # Present on the single-listing detail endpoint; the search list omits
    # these columns, so they default to None there.
    details: dict[str, Any] | None = None
    company_website: str | None = None


class SearchResponse(BaseModel):
    """Envelope for search results (server-side paginated)."""

    count: int  # total matching rows across ALL pages
    page: int
    page_size: int
    total_pages: int
    results: list[ListingOut]


class Suggestion(BaseModel):
    """One typeahead suggestion for the search box."""

    type: Literal["chemical", "supplier"]
    label: str  # what to display and drop into the search box when picked
    cas_number: str | None = None  # set for chemical suggestions when known
    count: int = 0  # how many listings back this suggestion


class SuggestResponse(BaseModel):
    """Envelope for the search typeahead suggestions."""

    suggestions: list[Suggestion]


_ALLOWED_SORTS = {"price_asc", "price_desc", "name_asc", "name_desc"}


class InterpretedFilters(BaseModel):
    """
    Structured filters the NL search agent extracted from a user's sentence.

    This is the ONLY thing the model produces — the result rows always come
    from database.search_listings (the same trusted query path as the manual
    filter search), never from the model itself.
    """

    name_query: str | None = None
    cas_number: str | None = None
    min_price: float | None = Field(default=None, ge=0)
    max_price: float | None = Field(default=None, ge=0)
    min_purity: float | None = Field(default=None, ge=0, le=100)
    max_purity: float | None = Field(default=None, ge=0, le=100)
    # Free-text match against the flexible listings.details JSONB (color,
    # hazard class, storage, ...) — best-effort, not a precision filter.
    details_query: str | None = None
    sort: str = "price_asc"

    @field_validator("name_query", "cas_number", "details_query", mode="before")
    @classmethod
    def _clean_text(cls, v: Any) -> str | None:
        if v is None:
            return None
        text = str(v).strip()
        return text[:100] or None

    @field_validator("sort", mode="before")
    @classmethod
    def _clean_sort(cls, v: Any) -> str:
        # A model that emits an unknown sort falls back to the default rather
        # than failing the whole request.
        return v if v in _ALLOWED_SORTS else "price_asc"


class AISearchRequest(BaseModel):
    """Body for POST /search/ai."""

    query: str = Field(min_length=1, max_length=300)


class AISearchResponse(SearchResponse):
    """AI search reply: what was understood + rows from the normal search path."""

    interpreted_filters: InterpretedFilters


class ListingUpdate(BaseModel):
    """
    Body for PATCH /listings/{id} (admin data-fix). All fields optional —
    only fields present in the request are changed; an explicit null clears
    a nullable field.
    """

    name_raw: str | None = None
    name_en: str | None = None
    cas_number: str | None = None
    price: float | None = Field(default=None, ge=0)
    currency: str | None = None
    purity: str | None = None
    needs_review: bool | None = None
    # Full replacement of the flexible technical-details object (the edit
    # form sends the whole edited object; the PubChem reference block is
    # preserved client-side). Does not affect the dedup key.
    details: dict[str, Any] | None = None
    # Reassign to an existing supplier (its id, from the suppliers directory).
    # Manual fix for when extraction found no supplier or the wrong one.
    company_id: int | None = None
    # Create a brand-new supplier with this name and assign it, when the
    # right one isn't in the directory yet. Mutually exclusive with
    # company_id in intent (the router prefers company_id if both are sent).
    new_company_name: str | None = None

    @field_validator("name_raw", "name_en")
    @classmethod
    def _names_non_empty(cls, v: str | None) -> str | None:
        # Names are required columns — an explicit null/blank would break rows.
        if v is not None and not v.strip():
            raise ValueError("must not be empty")
        return v.strip() if isinstance(v, str) else v

    @field_validator("new_company_name")
    @classmethod
    def _new_company_name_non_empty(cls, v: str | None) -> str | None:
        if v is not None and not v.strip():
            raise ValueError("must not be empty")
        return v.strip() if isinstance(v, str) else v


class BulkDeleteRequest(BaseModel):
    """Body for POST /listings/bulk-delete (admin bulk action)."""

    # Bounded so a runaway client can't ask for an unbounded delete; the UI
    # selects at most a page (15) at a time, so 200 is comfortable headroom.
    ids: list[str] = Field(min_length=1, max_length=200)


class BulkDeleteResult(BaseModel):
    """Result of POST /listings/bulk-delete."""

    deleted: int  # rows actually removed (stale/unknown ids are skipped)


class SupplierOut(BaseModel):
    """One supplier in the directory (GET /suppliers) — every extracted
    company detail plus derived listing count / printed websites."""

    id: int
    company_name: str = ""
    company_name_en: str | None = None
    email: str | None = None
    contact_number: str | None = None
    websites: list[str] = []  # distinct printed sites from this supplier's listings
    listing_count: int = 0
    created_at: str = ""  # first time a brochure introduced this supplier


class SuppliersResponse(BaseModel):
    """Envelope for the suppliers directory (server-side paginated)."""

    items: list[SupplierOut]
    count: int  # total suppliers across ALL pages
    page: int
    page_size: int
    total_pages: int


class ReviewItemOut(ListingOut):
    """A flagged listing in the review queue, annotated with the upload that
    produced it (null for uploads predating the listing_ids ledger column)."""

    source_content_hash: str | None = None
    source_filename: str | None = None


class ReviewQueueResponse(BaseModel):
    """Envelope for the needs-review queue (server-side paginated)."""

    items: list[ReviewItemOut]
    count: int  # total flagged listings across ALL pages
    page: int
    page_size: int
    total_pages: int


class AuditEntryOut(BaseModel):
    """One activity-log entry: who did what, when (GET /admin/audit)."""

    id: int
    actor: str | None = None  # user id
    actor_email: str | None = None  # resolved at read time via the Admin API
    action: str  # upload | update_listing | delete_listing | bulk_delete_listings | undo_upload
    details: dict[str, Any] = {}
    created_at: str


class AuditLogResponse(BaseModel):
    """Envelope for the activity log (server-side paginated)."""

    items: list[AuditEntryOut]
    count: int
    page: int
    page_size: int
    total_pages: int


class UndoUploadResult(BaseModel):
    """Result of DELETE /uploads/{content_hash} (undo one upload)."""

    deleted_listings: int
    # Listings kept because another upload also produced them (shared ids).
    kept_shared: int


class UploadJobOut(BaseModel):
    """One server-side upload job, as shown in the upload progress UI."""

    id: str
    filename: str
    status: Literal["queued", "processing", "paused", "done", "failed", "cancelled"]
    stage: str  # human-readable description of what is happening right now
    company_name: str | None = None
    products_found: int = 0
    listings_inserted: int = 0
    needs_review: int = 0
    duplicate: bool = False  # true when skipped as an already-processed PDF
    error: str | None = None
    created_at: float
    updated_at: float
    # Which controls are valid right now (drives the per-file buttons).
    can_pause: bool = False
    can_resume: bool = False
    can_cancel: bool = False
    can_restart: bool = False
    can_remove: bool = True


class UploadJobsResponse(BaseModel):
    """Envelope for the caller's upload jobs."""

    jobs: list[UploadJobOut]


class DocumentStatusOut(BaseModel):
    """One document in the new DB-worker pipeline, as shown in the upload
    progress UI. Progress is derived from the documents/pages status machine —
    there is no in-memory job any more."""

    id: str
    content_hash: str = ""
    filename: str = ""
    # splitting | split | extracting | done | failed (pending is transient).
    status: str
    page_count: int = 0
    pages_done: int = 0
    product_count: int = 0
    company_name: str | None = None
    duplicate: bool = False  # true when skipped as an already-processed PDF
    error: str | None = None
    created_at: str = ""


class DocumentStatusListResponse(BaseModel):
    """Envelope for the caller's recent pipeline documents (upload progress)."""

    documents: list[DocumentStatusOut]


class StartProcessingRequest(BaseModel):
    """Body for POST /upload-jobs/start. Omit `document_ids` (or the whole body)
    to release every staged upload the caller owns."""

    document_ids: list[str] | None = None


class StartProcessingResult(BaseModel):
    """How many staged uploads were released, and their updated status rows."""

    started: int
    documents: list[DocumentStatusOut]


class JobActionRequest(BaseModel):
    """Body for POST /upload-jobs/{id}/action."""

    action: Literal["pause", "resume", "cancel", "restart", "remove"]


class UploadHistoryItem(BaseModel):
    """One processed-upload record for the upload history / admin audit."""

    content_hash: str
    filename: str = ""
    company_name: str = ""
    company_name_en: str = ""
    product_count: int = 0
    uploaded_by: str | None = None
    uploaded_by_email: str | None = None  # populated only in the admin-wide view
    created_at: str


class UploadHistoryResponse(BaseModel):
    """Envelope for the upload history list (server-paginated)."""

    items: list[UploadHistoryItem]
    count: int = 0  # total matching uploads across ALL pages
    page: int = 1
    page_size: int = 10
    total_pages: int = 1


class DocumentListingsResponse(BaseModel):
    """The listings a single upload produced ("which ones were added")."""

    content_hash: str
    filename: str = ""
    listings: list[ListingOut]


class StatusDistribution(BaseModel):
    """The listing catalog split into three disjoint quality states, summing to
    total_listings. A listing flagged for review counts as needs_review even if
    it also lacks a price, so the three slices never double-count."""

    complete: int = 0
    needs_review: int = 0
    missing_price: int = 0


class SupplierListingCount(BaseModel):
    """One bar in the dashboard's 'Top suppliers by listings' panel."""

    company_id: int
    name: str
    count: int


class DashboardSummary(BaseModel):
    """Aggregate stat-card counts for the admin/manager dashboard."""

    total_listings: int = 0
    total_suppliers: int = 0
    uploads_last_7d: int = 0
    needs_review: int = 0
    status_distribution: StatusDistribution = StatusDistribution()
    top_suppliers: list[SupplierListingCount] = []


class ManagedUser(BaseModel):
    """One row on the admin user-management page."""

    id: str
    email: str | None = None
    role: str
    created_at: str | None = None
    last_sign_in_at: str | None = None


class UsersResponse(BaseModel):
    """Envelope for the admin user list."""

    users: list[ManagedUser]


class RoleUpdate(BaseModel):
    """Body for PATCH /users/{id}/role."""

    role: Literal["admin", "manager", "viewer"]
